accepted/wasmtime-debugging.md

```diff
@@ -0,0 +1,493 @@
+once we finally return from the tail callee.
+
+### Winch Considerations
+
+Interactive debugging support in an optimizing compiler like Cranelift would
+require fundamental changes to how optimizations are applied and how program
+state is tracked throughout compilation. Maintaining compiler performance while
+integrating those changes would be a significant undertaking. For this reason,
+for the foreseeable future, we are only going to support live debugging with
+Wasmtime’s baseline compiler: Winch.
+
+Winch will require some additional features to enable the debugging workflows
+outlined above:
+
+* The visitor that Winch uses to translate Wasm opcodes to native code will need
+  to be augmented to enable the interleaving of calls to local utility functions
+  that check for breakpoint/watchpoint status.
+
+* We will need to either maintain a mapping from locals to register/stack slot
+  for all Wasm PC points that we might hit a breakpoint or watchpoint, or force
+  winch to unconditionally spill locals to the stack. The latter would greatly
+  simplify tracking local locations, while the former would greatly increase the
+  amount of context we would need to pass into the utility function for
+  inspecting the execution state.
+
+# Future Work
+
+## Source Language Debugging
+
+While DWARF metadata is enough to enable debugging compiled languages, it
+doesn’t really help with debugging a running interpreter that has been compiled
+to Wasm — debugging that wasm module would amount to debugging the interpreter,
+not the underlying program being interpreted. Let’s examine this through two
+examples: debugging a Rust program compiled with DWARF annotations, and
+debugging a JavaScript programming being interpreted by SpiderMonkey compiled to
+Wasm.
+
+In the case of debugging compiled Rust that has DWARF annotations present, all
+of the existing workflows we described in this document would apply: source line
+mappings and the locations of local variables would all be determined by
+translating directly back to offsets in the wasm module. However if we’re
+debugging javascript running on SpiderMonkey compiled to Wasm, DWARF annotations
+would only give us information about the spidermonkey interpreter; we need to
+instruct the debugging infrastructure provided by SpiderMonkey to proceed with
+source-level debugging.
+
+The difference between these two workflows suggests that a debugger
+implementation in Wasmtime needs to understand the target being debugged at a
+deeper level than just annotations. To address this problem, we plan to enrich
+the debug adapter interface described for the null and DWARF adapters above with
+enough information to allow control of a guest interpreter. This will include
+the need to reflect on symbols exported by the interpreter allowing for
+functions to be dynamically called by the debug adapter, reusing all of the
+interpreter’s existing debugging facilities directly.
```

Comment:
One of the ways I thought about doing this is providing support for nested DAPs; SpiderMonkey for example would be compiled with a DAP implementation that proxies DAP requests to its own debugging facilities, and we'd be able to proxy the inner DAP it back to the one exposed by the host DAP server.